### DISCUSSION TITLE
Build order part location export

### DISCUSSION THREAD

**Discussion by martin323232** [inventree/InvenTree#4029] [Q&A]
Is it possible export allocated part locations for a build order?
I can see location when pressing the "plus" icon and manually writing them down, but an export to pdf with part name and location would be nice.

![image](https://user-images.githubusercontent.com/6439801/206388215-83723627-4240-4e13-b819-2f860cd3bc33.png)

**Comment by SergeoLacruz** [comment]
Hi,
This would be really interesting. I am already looking for a way to get these parts in the reporting system but i did not find the proper 
context variables yet. 
I tried .sub_part.build_order_allocations and .required_parts.build_order_allocations but as these variables are related to the part  the resulting query set lists all allocations, also the ones from other BOs. Filtering those might get too complex. If I had the required variables I would like to create something like a packing list. See you,

Michael

**Reply by SchrodingersGat** [reply]
The individual stock items allocations are available as the variable `build.allocated_stock`:

https://github.com/inventree/InvenTree/blob/cd29ddaf123b686d73e5e4308520e7b46373be22/InvenTree/build/models.py#L1328

Note that these are `BuildItem` instances:

https://github.com/inventree/InvenTree/blob/cd29ddaf123b686d73e5e4308520e7b46373be22/InvenTree/build/models.py#L1127

**Reply by SergeoLacruz** [reply]
Hi,
Thanks for the info. So using  the code:
```python
  <table class='changes-table'>
    <thead>
      <tr>
        <th>Original IPN</th>
        <th>Reference</th>
        <th>Allocated Part</th>
        <th>Location</th>
      </tr>
    </thead>
    <tbody>
  {% for line in build.allocated_stock.all %}
      <tr>
        <td> {{ line.bom_item.sub_part.IPN }} </td>
        <td> {{ line.bom_item.reference }} </td>
        {% if line.stock_item.part.IPN == "DNP-000000-000" %}
        <td> do not place </td>
        <td> </td>
        {% else %}
        <td> {{ line.stock_item.part.IPN }} </td>
        <td> {{ line.stock_item.location.name }} </td>
        {% endif %}
      </tr>
  {% endfor %}
    </tbody>
  </table>

```
I get this:

| Original IPN | Reference | Allocated Part | Location |
| ----- | ------ | ------- | ------------------|
| TRA-576006-001 | T1, T2 | TRA-576006-001 | 17 |
| CAP-136006-001 | C2, C4, C5 | CAP-136006-002 | 06 |
| ANA-238003-001 | IC1 | ANA-238003-001 | 02|
| RES-045016-001 | R1, R2 | RES-045012-001 | 27 |
| TRA-580008-001 | T100, T101 | do not place |  


That looks already good. 
Here I look at the allocated_stock and get from this the part that war originally there using " line.bom_item.sub_part.IPN". 
I would prefer the other direction.  This would also list parts where the allocation is missing. 

Michael

**Reply by martin323232** [reply]
Thanks Michael for the code, this was just what I was looking for. 

I’ve added a column for manufacturer part number, also using this on our labels. 
The only way I’ve managed to use this is with the variable “item.supplier_part.manufacturer_part” but this also outputs manufacturer name, “Samsung Electro-Mechanics | CL10A475KQ8NNNC”. 
How can I just get the manufacturer part name?

**Reply by SergeoLacruz** [reply]
Just guessing: manufacturer_part.name?

**Reply by martin323232** [reply]
There is no output when using manufacturer_part.name.
would also like to add part description. In labels "part.description" works, but no output in Build Order, {{ line.part.description }}

**Reply by SchrodingersGat** [reply]
If you are looking for the "manufacturer part name" you should be using `item.supplier_part.manufacturer_part.MPN`

You can make use of any attribute which exists on the particular model:

https://github.com/inventree/InvenTree/blob/55c14bd2386a8b26a3a0f3e5183e5519a95ff6d1/InvenTree/company/models.py#L272

**Reply by martin323232** [reply]
thanks @SchrodingersGat 

item.supplier_part.manufacturer_part.MPN worked for stock label, but no output in BO report

` <td> {{ line.bom_item.reference }} </td>
		<td> {{ line.item.supplier_part.manufacturer_part.MPN  }} </td>		
        {% if line.stock_item.part.IPN == "DNP-000000-000" %}
        <td> do not place </td>
        <td> </td>
        {% else %}
        <td> {{ line.part.description }} </td>
        <td> {{ line.stock_item.location.name }} </td>`

**Comment by SergeoLacruz** [comment]
manufacturer_parts and supplier_parts can be several. So try e.g.
```python 
  <table class='changes-table'>
    <thead>
      <tr>
        <th>.stock_item.part.IPN</th>
        <th>.stock_item.part.supplier_parts.all.0.SKU</th>
        <th>.stock_item.part.manufacturer_parts.all.0.MPN</th>
      </tr>
    </thead>
    <tbody>
  {% for line in build.allocated_stock.all %}
      <tr>
        <td> {{ line.stock_item.part.IPN }} </td>
        <td> {{ line.stock_item.part.supplier_parts.all.0.SKU }} </td>
        <td> {{ line.stock_item.part.manufacturer_parts.all.0.MPN }} </td>
      </tr>
  {% endfor %}
    </tbody>
  </table>
```
With .0 you get just the first from the list. 01 will sow the second ...

**Reply by martin323232** [reply]
@SergeoLacruz Thanks that worked perfectly.

**Comment by SergeoLacruz** [comment]
Hi,
You are welcome :-)
I am still locking for an easy way to find out if a build.bom_item.x has an allocated stock for that  build or not. Anybody any idea?
Michael
